File: frontend/demo/component/datepicker/date-picker-custom-validation.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/date-picker';
import { html, LitElement } from 'lit';
import { customElement } from 'lit/decorators.js';
import { Binder, field } from '@vaadin/hilla-lit-form';
import AppointmentModel from 'Frontend/generated/com/vaadin/demo/domain/AppointmentModel';
import { applyTheme } from 'Frontend/demo/theme';

@customElement('date-picker-custom-validation')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    applyTheme(root);
    return root;
  }

  // tag::snippet[]
  private binder = new Binder(this, AppointmentModel);

  protected override firstUpdated() {
    this.binder.for(this.binder.model.startDate).addValidator({
      message: 'Select a weekday',
      validate: (startDate: string) => {
        const date = new Date(startDate);
        const isWeekday = date.getDay() >= 1 && date.getDay() <= 5;
        return isWeekday;
      },
    });
  }

  protected override render() {
    return html`
      <vaadin-date-picker
        label="Meeting date"
        helper-text="Mondays – Fridays only"
        ${field(this.binder.model.startDate)}
      ></vaadin-date-picker>
    `;
  }
  // end::snippet[]
}

File: src/main/java/com/vaadin/demo/component/datepicker/DatePickerCustomValidation.java
package com.vaadin.demo.component.datepicker;

import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.demo.domain.Appointment;
import com.vaadin.flow.component.datepicker.DatePicker;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.data.binder.Binder;
import com.vaadin.flow.router.Route;

@Route("date-picker-custom-validation")
public class DatePickerCustomValidation extends Div {

    public DatePickerCustomValidation() {
        // tag::snippet[]
        DatePicker datePicker = new DatePicker("Meeting date");
        datePicker.setHelperText("Mondays – Fridays only");

        Binder<Appointment> binder = new Binder<>(Appointment.class);
        binder.forField(datePicker)
                .withValidator(localDate -> {
                    int dayOfWeek = localDate.getDayOfWeek().getValue();
                    boolean validWeekDay = dayOfWeek >= 1 && dayOfWeek <= 5;
                    return validWeekDay;
                }, "Please select a weekday")
                .bind(Appointment::getStartDate, Appointment::setStartDate);
        // end::snippet[]

        add(datePicker);
    }
    public static class Exporter extends DemoExporter<DatePickerCustomValidation> { // hidden-source-line
    } // hidden-source-line
}
